src/method.jl

```diff
@@ -58,24 +58,61 @@ function argument_types(sig)
     return parameters(sig)[2:end]
 end
 
-name_of_type(x) = x
+module DummyThatHasOnlyDefaultImports end  # for working out visability
+
+function name_of_module(m::Module)
+    if Base.is_root_module(m)
+        return nameof(m)
+    else
+        return :($(name_of_module(parentmodule(m))).$(nameof(m)))
+    end
+end
+function name_of_type(x::Core.TypeName)
+    #TODO: could let user pass this in, then we could be using what is inscope for them
+    # but this is not important as we will give a correct (if overly verbose) output as is.
+    from = DummyThatHasOnlyDefaultImports
+    if Base.isvisible(x.name, x.module, from)  # avoid qualifying things that are in scope
+        return x.name
+    else
+        return :($(name_of_module(x.module)).$(x.name))
+    end
+end
+
+name_of_type(x::Symbol) = x  # Literal
+function name_of_type(x::T) where T  # Literal
+    # it is a bug in our implementation if this error every gets hit.
+    isbits(x) || throw(DomainError((x, T), "not a valid type-param"))
+    return x
+end
```

Comment:
Added these literals rather than `name_of_type(x)` so i can be sure nothing slipped through